docs/SyntaxForModels.mdx

```diff
@@ -111,6 +111,112 @@ The above matcher is the simplest, it means that the subject, object and action
 
 You can use arithmetic like ``+, -, *, /`` and logical operators like ``&&, ||, !`` in matchers.
 
+### Orders of expressions in matchers
+The order of expressions can greatly affect performance.  
+Look at the following example for details:  
+
+```go
+const rbac_models = `
+[request_definition]
+r = sub, obj, act
+
+[policy_definition]
+p = sub, obj, act
+
+[role_definition]
+g = _, _
+
+[policy_effect]
+e = some(where (p.eft == allow))
+
+[matchers]
+m = g(r.sub, p.sub) && r.obj == p.obj && r.act == p.act
+`
+
+func TestManyRoles(t *testing.T) {
+
+	m, _ := model.NewModelFromString(rbac_models)
+	e, _ := NewEnforcer(m, false)
+
+	roles := []string{"admin", "manager", "developer", "tester"}
+
+	// 2500 projects
+	for nbPrj := 1; nbPrj < 2500; nbPrj++ {
+		// 4 objects and 1 role per object (so 4 roles)
+		for _, role := range roles {
+			roleDB := fmt.Sprintf("%s_project:%d", role, nbPrj)
+			objectDB := fmt.Sprintf("/projects/%d", nbPrj)
+			e.AddPolicy(roleDB, objectDB, "GET")
+		}
+		jasmineRole := fmt.Sprintf("%s_project:%d", roles[1], nbPrj)
+		e.AddGroupingPolicy("jasmine", jasmineRole)
+	}
+
+	e.AddGroupingPolicy("abu", "manager_project:1")
+	e.AddGroupingPolicy("abu", "manager_project:2499")
+
+	// With same number of policies
+	// 	User 'abu' has only two roles
+	// 	User 'jasmine' has many roles (1 role per policy, here 2500 roles)
+
+	request := func(subject, object, action string) {
+		t0 := time.Now()
+		resp, _ := e.Enforce(subject, object, action)
+		tElapse := time.Since(t0)
+		t.Logf("RESPONSE %-10s %s\t %s : %5v IN: %+v", subject, object, action, resp, tElapse)
+		if tElapse > time.Millisecond*100 {
+			t.Errorf("More than 100 milliseconds for %s %s %s : %+v", subject, object, action, tElapse)
+		}
+	}
+
+	request("abu", "/projects/1", "GET")        // really fast because only 2 roles in all policies and at the beginning of the casbin_rule table
+	request("abu", "/projects/2499", "GET")     // fast because only 2 roles in all policies
+	request("jasmine", "/projects/1", "GET")    // really fast at the beginning of the casbin_rule table
+
+	request("jasmine", "/projects/2499", "GET") // slow and fail the only 1st time   <<<< pb here
+	request("jasmine", "/projects/2499", "GET") // fast maybe due to internal cache mechanism
+	
+        // same issue with non-existing roles
```

Comment:
![image](https://user-images.githubusercontent.com/53544726/210314530-275ef2d3-3a18-4d36-9f47-62c6f9ca3659.png)

please format this line

Reply:
Thanks, fixed it already.

Reply:
Please do not mix tabs and spaces

![image](https://user-images.githubusercontent.com/53544726/210319035-62efbc39-84f3-4b07-a876-d51cdd9b37ab.png)

You can use vscode (or other editor) to convert all tab to spaces

![image](https://user-images.githubusercontent.com/53544726/210319374-01131391-3de5-4b7b-86f4-9d54d4ae6854.png)

![image](https://user-images.githubusercontent.com/53544726/210319190-c898697b-a9ef-41d0-98c9-467e86993252.png)

Reply:
Sorry for that. I just copy the code from the author.
Fixed it.